src/browser-polyfill.js

```diff
@@ -361,18 +361,44 @@ if (typeof browser === "undefined") {
        *        yield a response. False otherwise.
        */
       return function onMessage(message, sender, sendResponse) {
-        let result = listener(message, sender);
+        let didCallSendResponse = false;
+
+        let wrappedSendResponse;
+        let sendResponsePromise = new Promise(resolve => {
+          wrappedSendResponse = function(response) {
+            didCallSendResponse = true;
+            resolve(response);
+          };
+        });
+
+        let result = listener(message, sender, wrappedSendResponse);
+
+        const isResultThenable = result !== true && isThenable(result);
 
-        if (isThenable(result)) {
+        // If the listener didn't returned true or a Promise, or called
+        // wrappedSendResponse synchronously, we can exit earlier
+        // because there will be no response sent from this listener.
+        if (result !== true && !isResultThenable && !didCallSendResponse) {
+          return false;
+        }
+
+        // If the listener returned a Promise, send the resolved value as a
+        // result, otherwise wait the promise related to the wrappedSendResponse
+        // callback to resolve and send it as a response.
+        if (isResultThenable) {
           result.then(sendResponse, error => {
             console.error(error);
             sendResponse(error);
-          });
-
-          return true;
-        } else if (result !== undefined) {
-          sendResponse(result);
+          }).catch(err => console.error);
```

Comment:
(also on line 397)
`catch(console.error)` or `err => console.error(err)` . Not `err => console.error`.

Chrome logs uncaught promise rejections to the console, so you don't even need to call `console.error` here.


Side note: The polyfill is quite chatty. Chrome is generally quiet about messages in `onMessage` / `sendMessage`, unless an obviously incorrect error occurs (e.g. calling `sendResponse` without `return true;`). Firefox doesn't print rejected promises either, it just forwards the promise result to the `sendMessage` caller.

An error is usually an indication that something is wrong, so I can see value in preserving the `console.error` call. To not change the behavior too much,  

Usually error messages are not even serializable (they become `{}` in Chrome), so perhaps it would be better to just invoke `sendResponse` without any value (`undefined`).